File: android/app/src/main/java/com/bitcraftapps/stax/data/billing/EntitlementManager.kt
package com.bitcraftapps.stax.data.billing

import android.content.Context
import androidx.compose.runtime.compositionLocalOf
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

val LocalEntitlementManager = compositionLocalOf<EntitlementManager> {
    error("No EntitlementManager provided")
}

class EntitlementManager(context: Context) {

    private val prefs = context.getSharedPreferences("stax_subscription", Context.MODE_PRIVATE)

    private val _subscriptionState = MutableStateFlow(readStateFromPrefs())
    val subscriptionState: StateFlow<SubscriptionState> = _subscriptionState.asStateFlow()

    private val _isPremium = MutableStateFlow(_subscriptionState.value is SubscriptionState.Premium)
    val isPremium: StateFlow<Boolean> = _isPremium.asStateFlow()

    private fun readStateFromPrefs(): SubscriptionState {
        val stateStr = prefs.getString("subscription_state", "free") ?: "free"
        return when (stateStr) {
            "premium" -> {
                val isInTrial = prefs.getBoolean("is_in_trial", false)
                val expiryMs = prefs.getLong("expiry_ms", Long.MAX_VALUE)
                SubscriptionState.Premium(isInTrial, expiryMs)
            }
            "expired" -> {
                val gracePeriodEndsMs = prefs.getLong("grace_period_ends_ms", -1L)
                SubscriptionState.Expired(if (gracePeriodEndsMs == -1L) null else gracePeriodEndsMs)
            }
            else -> SubscriptionState.Free
        }
    }

    private fun emitState(state: SubscriptionState) {
        _subscriptionState.value = state
        _isPremium.value = state is SubscriptionState.Premium
    }

    fun setPremium(isInTrial: Boolean, expiryMs: Long) {
        prefs.edit()
            .putString("subscription_state", "premium")
            .putBoolean("is_in_trial", isInTrial)
            .putLong("expiry_ms", expiryMs)
            .apply()
        emitState(SubscriptionState.Premium(isInTrial, expiryMs))
    }

    fun setExpired(gracePeriodEndsMs: Long?) {
        prefs.edit()
            .putString("subscription_state", "expired")
            .putLong("grace_period_ends_ms", gracePeriodEndsMs ?: -1L)
            .apply()
        emitState(SubscriptionState.Expired(gracePeriodEndsMs))
    }

    fun setFree() {
        prefs.edit()
            .putString("subscription_state", "free")
            .remove("is_in_trial")
            .remove("expiry_ms")
            .remove("grace_period_ends_ms")
            .apply()
        emitState(SubscriptionState.Free)
    }

    fun recordScan() {
        val key = scanKeyForToday()
        val current = prefs.getInt(key, 0)
        prefs.edit().putInt(key, current + 1).apply()
    }

    fun getDailyScans(): Int = prefs.getInt(scanKeyForToday(), 0)

    private fun scanKeyForToday(): String {
        val dateStr = SimpleDateFormat("yyyy-MM-dd", Locale.US).format(Date())
        return "scans_$dateStr"
    }

    fun checkLimit(
        feature: Feature,
        sessionPhotoCount: Int? = null,
        totalSessions: Int? = null,
        favoritesCount: Int? = null,
        casinoIndex: Int? = null
    ): LimitResult {
        if (_isPremium.value) return LimitResult.Allowed

        return when (feature) {
            Feature.SCAN -> {
                val scans = getDailyScans()
                if (scans < MAX_FREE_SCANS_PER_DAY) LimitResult.Allowed
                else LimitResult.Blocked(Feature.SCAN)
            }
            Feature.AI_SCAN -> LimitResult.Blocked(Feature.AI_SCAN)
            Feature.SESSION_CREATE -> {
                val count = totalSessions ?: 0
                if (count < MAX_FREE_SESSIONS) LimitResult.Allowed
                else LimitResult.Blocked(Feature.SESSION_CREATE)
            }
            Feature.PHOTO_ADD -> {
                val count = sessionPhotoCount ?: 0
                if (count < MAX_FREE_PHOTOS_PER_SESSION) LimitResult.Allowed
                else LimitResult.Blocked(Feature.PHOTO_ADD)
            }
            Feature.CHIP_CONFIG -> {
                val idx = casinoIndex ?: 0
                if (idx < MAX_FREE_CHIP_CONFIG_CASINOS) LimitResult.Allowed
                else LimitResult.Blocked(Feature.CHIP_CONFIG)
            }
            Feature.FAVORITES -> {
                val count = favoritesCount ?: 0
                if (count < MAX_FREE_FAVORITES) LimitResult.Allowed
                else LimitResult.Blocked(Feature.FAVORITES)
            }
            Feature.SHARE_CLEAN -> LimitResult.Blocked(Feature.SHARE_CLEAN)
        }
    }

    fun getTrialDaysRemaining(): Int {
        val state = _subscriptionState.value
        if (state !is SubscriptionState.Premium || !state.isInTrial) return 0
        val remainingMs = state.expiryMs - System.currentTimeMillis()
        return TimeUnit.MILLISECONDS.toDays(remainingMs).toInt().coerceAtLeast(0)
    }
}

File: ios/Stax/Billing/EntitlementManager.swift
import Foundation
import StoreKit

@MainActor
class EntitlementManager: ObservableObject {

    @Published var isPremium: Bool = false
    @Published var isInTrial: Bool = false
    @Published var subscriptionState: SubscriptionState = .free

    private var cachedExpiryDate: Date = .distantFuture

    // MARK: – Scan counter (keyed by date)

    private func todayKey() -> String {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd"
        return "scans_\(fmt.string(from: Date()))"
    }

    func recordScan() {
        let key = todayKey()
        let current = UserDefaults.standard.integer(forKey: key)
        UserDefaults.standard.set(current + 1, forKey: key)
    }

    func getDailyScans() -> Int {
        UserDefaults.standard.integer(forKey: todayKey())
    }

    // MARK: – State mutations

    func update(from transaction: Transaction) {
        let inTrial = transaction.offerType == .introductoryOffer
        let expiry = transaction.expirationDate ?? .distantFuture
        isInTrial = inTrial
        cachedExpiryDate = expiry
        subscriptionState = .premium(isInTrial: inTrial, expiryDate: expiry)
        isPremium = true
    }

    func setExpired() {
        subscriptionState = .expired
        isPremium = false
        isInTrial = false
    }

    func setFree() {
        subscriptionState = .free
        isPremium = false
        isInTrial = false
    }

    // MARK: – Limit checking

    func checkLimit(
        for feature: Feature,
        sessionPhotoCount: Int? = nil,
        totalSessions: Int? = nil,
        favoritesCount: Int? = nil,
        casinoIndex: Int? = nil
    ) -> LimitResult {
        if isPremium { return .allowed }

        switch feature {
        case .scan:
            if getDailyScans() < FreeTierLimits.maxScansPerDay { return .allowed }
            return .blocked(.scan)

        case .aiScan:
            return .blocked(.aiScan)

        case .sessionCreate:
            let count = totalSessions ?? 0
            if count < FreeTierLimits.maxSessions { return .allowed }
            return .blocked(.sessionCreate)

        case .photoAdd:
            let count = sessionPhotoCount ?? 0
            if count < FreeTierLimits.maxPhotosPerSession { return .allowed }
            return .blocked(.photoAdd)

        case .chipConfig:
            let index = casinoIndex ?? 0
            if index < FreeTierLimits.maxChipConfigCasinos { return .allowed }
            return .blocked(.chipConfig)

        case .favorites:
            let count = favoritesCount ?? 0
            if count < FreeTierLimits.maxFavorites { return .allowed }
            return .blocked(.favorites)

        case .shareClean:
            return .blocked(.shareClean)
        }
    }

    #if DEBUG
    func setDebugPremium() {
        let expiry = Calendar.current.date(byAdding: .year, value: 1, to: Date()) ?? .distantFuture
        cachedExpiryDate = expiry
        subscriptionState = .premium(isInTrial: false, expiryDate: expiry)
        isPremium = true
        isInTrial = false
    }
    #endif

    // MARK: – Trial countdown

    func getTrialDaysRemaining() -> Int {
        guard case .premium(let inTrial, let expiry) = subscriptionState, inTrial else { return 0 }
        let days = Calendar.current.dateComponents([.day], from: Date(), to: expiry).day ?? 0
        return max(0, days)
    }
}
